src/hotspot/share/runtime/objectMonitor.cpp

```diff
@@ -402,29 +402,34 @@ bool ObjectMonitor::enter(JavaThread* current) {
     }
 
     OSThreadContendState osts(current->osthread());
-    ThreadBlockInVM tbivm(current);
 
-    // TODO-FIXME: change the following for(;;) loop to straight-line code.
-    for (;;) {
-      current->set_suspend_equivalent();
-      // cleared by handle_special_suspend_equivalent_condition()
-      // or java_suspend_self()
+    assert(current->thread_state() == _thread_in_vm, "invariant");
 
+    current->frame_anchor()->make_walkable(current);
+    OrderAccess::storestore();
+    for (;;) {
+      current->set_thread_state(_thread_blocked);
       EnterI(current);
-
-      if (!current->handle_special_suspend_equivalent_condition()) break;
-
-      // We have acquired the contended monitor, but while we were
-      // waiting another thread suspended us. We don't want to enter
-      // the monitor while suspended because that would surprise the
-      // thread that suspended us.
-      //
-      _recursions = 0;
-      _succ = NULL;
-      exit(false, current);
-
-      current->java_suspend_self();
+      current->set_thread_state_fence(_thread_blocked_trans);
+      if (SafepointMechanism::should_process(current) &&
+        current->suspend_request_pending()) {
```

Comment:
We don't want to always to take a lock (handshake state lock), if the poll is disarmed there cannot be any suspend request.
We only need to unlock the OM for suspends that happened before we grabbed the OM lock in EnterI().

When we leave EnterI we are still blocked, this means a thread might be processing the suspend handshake, which was emitted before we did EnterI().
To synchronize the suspend check with such a thread we need to grab the HandshakeState lock, e.g. we cannot racy check the suspended flag.

I choosed to check for an async suspension handshake that needs to be processed. (We can have such without being suspended). We could ignored that async handshake by just checking is_suspended, it would be processed in the else statement instead with "SafepointMechanism::process_if_requested(current);" instead.

But we avoid releasing the OM lock in cases where we already have been resumed.

So I can change to is_suspended inside the the method, but we still must do it under the HandshakeState lock.

Reply:
To clarify, is_suspended() can only be checked when the JavaThread is _unsafe_, *after* you have transitioned from the safe state.
In this case we are checking if we are suspended before we have completed the transition because we need to know if we should drop the OM lock.

Reply:
> 
> 
> We don't want to always to take a lock (handshake state lock), if the poll is disarmed there cannot be any suspend request.
> We only need to unlock the OM for suspends that happened before we grabbed the OM lock in EnterI().
> 
> When we leave EnterI we are still blocked, this means a thread might be processing the suspend handshake, which was emitted before we did EnterI().
> To synchronize the suspend check with such a thread we need to grab the HandshakeState lock, e.g. we cannot racy check the suspended flag.

I don't think that checking the suspended flag without holding the handshake
state lock would be too racy. Holding the OM is sufficient synchronization for
checking the suspended flag here.

If the suspender thread S did the suspend request while holding the OM then the
current thread T will see that it was suspended when it has entered the OM.
If S did the suspend without holding OM, only then the check would be racy but that would be ok.

> I choosed to check for an async suspension handshake that needs to be processed. (We can have such without being suspended). We could ignored that async handshake by just checking is_suspended, it would be processed in the else statement instead with "SafepointMechanism::process_if_requested(current);" instead.
> 
> But we avoid releasing the OM lock in cases where we already have been resumed.
> 
> So I can change to is_suspended inside the the method, but we still must do it under the HandshakeState lock.

Reply:
> 
> 
> To clarify, is_suspended() can only be checked when the JavaThread is _unsafe_, _after_ you have transitioned from the safe state.

For a minute I misunderstood this and thought you meant this as a general rule
for calling `is_suspended()` while there are examples where it is called without
any synchronization (e.g. in `JvmtiEnv::GetThreadState()`) which can be ok.

> In this case we are checking if we are suspended before we have completed the transition because we need to know if we should drop the OM lock.

I think it is sufficient to hold the OM when calling `is_suspended` to decide if OM has to be dropped.

Reply:
I do not follow, the OM is unrelated.
The suspender do not hold the OM.

What happens is:
- Thread A wait on OM X in blocked.
- Thread Z suspends Thread A, thread Z have no relation to OM X.
- Thread B unlocks OM X, thread B have no relation to the suspend request.
- Thread A locks OM X while blocked.
- Thread A was not allowed to lock OM X due to it is actually suspended.
- Thread A unlocks OM X and waits until resumed.

So while A and B fight over OM X, A is trying to figure out if Z suspended him while grabbing the lock.

The suspend flag and 'async handshake' flag are protected by the HandshakeState lock. (this is not lockless protocol)
The stores in SuspendThreadHandshake are not ordered, so they may happen in any order.
To read the stores there you must hold the HandshakeState lock.
Therefore if we want the exact answer to am A suspended we must read the flag under the lock.

Makes sense?

Reply:
> 
> 
> I do not follow, the OM is unrelated.
> The suspender do not hold the OM.
> 
> What happens is:
> 
>     * Thread A wait on OM X in blocked.
> 
>     * Thread Z suspends Thread A, thread Z have no relation to OM X.
> 
>     * Thread B unlocks OM X, thread B have no relation to the suspend request.
> 
>     * Thread A locks OM X while blocked.
> 
>     * Thread A was not allowed to lock OM X due to it is actually suspended.
> 
>     * Thread A unlocks OM X and waits until resumed.
> 
> 
> So while A and B fight over OM X, A is trying to figure out if Z suspended him while grabbing the lock.

If understand the example correctly then the suspend operation and the
operations on OM X are unordered with respect to each other. If so, then we can
put them into an order where the suspend happens after "Thread A locks OM X while
blocked.". We are free to do this; there's no synchronization that would prevent it.

Also when the suspend request happened while A was blocked then after
`current->set_thread_state_fence(_thread_blocked_trans);` the check of
`is_suspended()` will return true even if the handshake state lock is not
acquired for the check. And if Z tried to suspend A after the state change then
Z will block because A is not safe anymore.

Sorry for insisting. I really hope I'm not wrong ;)

Reply:
The only reason _suspended is volatile is to be able to make the the fast check in resume().
So disregard that early check and that it is volatile, the users of the flag uses HandshakeState lock for synchronizing reads and stores to that flag.

E.g
`set_suspend(true);
  set_async_suspend_handshake(true);
  log_trace(thread, suspend)("JavaThread:" INTPTR_FORMAT " suspended, arming ThreadSuspension", p2i(_handshakee));
  ThreadSuspensionHandshake* ts = new ThreadSuspensionHandshake();
  Handshake::execute(ts, _handshakee);
`
Before we have enqueued the async handshake we may have the poll disarmed and but the flag set_suspend(true).
This means we will miss the async handshake and continue to executed with suspended flag.

Both flags and async handshake are set atomically by serializing over the HnadshakeState lock.

In this case we want to both know if we are suspended if so process the suspension.

Reply:
(technically this will be ordered by the poll is since polls are only disarmed by threads them selfs)

Reply:
> Also when the suspend request happened while A was blocked then after
> current->set_thread_state_fence(_thread_blocked_trans); the check of
> is_suspended() will return true even if the handshake state lock is not
> acquired for the check

That statement of mine is wrong. (Hopefully) correct is that after the complete
state change from _thread_blocked to _thread_in_vm which includes being blocked
for a safepoint/handshake the current thread would be able to check
is_suspended() without holding the handshake state lock. It does not make a lot
of sense then because in an unsafe state JavaThread::current()->is_suspended()
will always return false.

> Thread A wait on OM X in blocked.
> Thread Z suspends Thread A, thread Z have no relation to OM X.
> Thread B unlocks OM X, thread B have no relation to the suspend request.
> Thread A locks OM X while blocked.
> Thread A was not allowed to lock OM X due to it is actually suspended.
> Thread A unlocks OM X and waits until resumed.

I understand that example now too: OM and suspend operations are unrelated. So I
thought it would be ok for A to enter OM X, but it is not. A thread must not
leave a safe state if it was suspended in that state (with a handshake). If it
did, e.g., its stack could become not walkable. And it must not enter the
monitor either.

Sorry for the confusion. The check is good.

Reply:
But I still think we can just call is_suspended() here. If is_suspended() returns false, then Z hasn't finished suspending A yet (either async operation was still not added to the queue in case this is the first suspend for A, or do_handshake() hasn't yet finished for that SuspendThreadHandshake operation in case there was already an async operation in the queue). Since we have already acquired OM, then Z will not see that after being suspended A switched from owning OM to not owning it. It will always see that A owned OM. And A will still block for the suspension in the 'else' clause since poll is armed due to SuspendThreadHandshake operation(even if async operation was still not added to the queue at the time A called is_suspended()), but we wouldn't need to release OM.
If is_suspended() returns true then at most we will have a false positive, since it could be that after the check somebody already resumed A, but that can also happen even with suspend_request_pending() right after releasing HandshakeState::_lock.
Is that correct or am I missing something?

Reply:
I'm not saying it doesn't work, I'm saying it works by accident (I never intended to be able to read the flag without HandshakeState lock).
So yes, if the flag is false, it is not possible for the suspend to have happened before we locked the OM.
And once poll is armed for the synchrone handshake it will stay armed until after we executed the asynchrone handshake trap.

Reply:
> But I still think we can just call is_suspended() here

@pchilano, potentially I had similar thoughts when I asked about it before the long weekend: the current thread will suspend either way in the then or else block if the execution of the SuspendThreadHandshake was started while the current thread was _thread_blocked. If the suspend is ordered with operations on OM then is_suspended() will return true. Otherwise it is ok to become the owner. This is a more intuitive reasoning. I prefer yours.

While looking at this I noticed that suspending in L428 could be problematic because the OM is just entered half way.

Maybe we could use plain ThreadBlockInVM and deal with unlocking the OM very late just before waiting in HandshakeState::suspend_in_handshake().